### PR TITLE
Add mentoring notes for Rust Circular Buffer exercise

### PR DESCRIPTION
**Pull request by seanchen1991**
This PR adds a new set of mentoring notes for the Circular Buffer exercise with multiple example solutions and common suggestions.

### CODE REVIEW

File: tracks/rust/exercises/circular-buffer/mentoring.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    .map(|element| {
+      self.update_read_index();
+      element
+    })
+}
+```
+
+While the first implementation is unlikely to ever actually panic due to the `unwrap` call due to how it's structured, imparting strategies for how to limit usage of `unwrap` will certainly be beneficial to the student(s). 
+
+- Logically distinct operations such as advancing the read index and advancing the write index should be relegated to their own functions to help with code readability, organization, and to adhere to the DRY principle.
```
Comment (SleeplessByte) [inline tracks/rust/exercises/circular-buffer/mentoring.md]: List is not correctly formatted (all two spaces), but I would suggest not using a list at all but add subheaders.

File: tracks/rust/exercises/circular-buffer/mentoring.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+- [Defining Error Types](https://doc.rust-lang.org/rust-by-example/error/multiple_error_types/define_error_type.html)
+- [Trait Bounds](https://doc.rust-lang.org/rust-by-example/generics/bounds.html)
+
+### Reasonable Solutions
+
+A reasonable solution should:
+
+- Return an `Error` when appropriate
+- Limit the use of `unwrap`
+- Exhibit Separation of Concerns by defining additional helper functions to perform logically distinct operations
```
Comment (SleeplessByte) [inline tracks/rust/exercises/circular-buffer/mentoring.md]: Optionally add a link to "Separation of Concerns" for those mentors who are less-acquainted.

File: tracks/rust/exercises/circular-buffer/mentoring.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        self.end = (self.end + 1) % self.buffer.len();
+    }
+}
+
+```
+
+### Common Suggestions
+
+- Usage of `unwrap` should be as minimal as possible, preferably non-existent if we can help it. Take the following implementation of the `read` function that calls `unwrap`:
+```rust
```
Comment (SleeplessByte) [inline tracks/rust/exercises/circular-buffer/mentoring.md]: Add whitespace (empty line) for consistency with rest of file.

```suggestion

```rust
```

File: tracks/rust/exercises/circular-buffer/mentoring.md [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+      self.update_read_index();
+      Ok(self.buffer[old_index].take().unwrap())
+    },
+    None => Err(Error::EmptyBuffer),
+  }
+}
+```
+
+How can we eliminate the `unwrap` call on `take`? The handy `ok_or` method is a prime candidate. This implementation can be restructured into something like this:
+```rust
```
Comment (SleeplessByte) [inline tracks/rust/exercises/circular-buffer/mentoring.md]:
```suggestion

```rust
```

Add whitespace (empty line) for consistency with rest of file.

File: tracks/rust/exercises/circular-buffer/mentoring.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    fn advance_end(&mut self) {
+        self.end = (self.end + 1) % self.buffer.len();
+    }
+}
+
+```
+
+### Common Suggestions
+
+- Usage of `unwrap` should be as minimal as possible, preferably non-existent if we can help it. Take the following implementation of the `read` function that calls `unwrap`:
```
Comment (SleeplessByte) [inline tracks/rust/exercises/circular-buffer/mentoring.md]: See comment lower about usage of list. It doesn't translate well with "mixed content" like this. I would use sub headers here.

**Comment by workingjubilee** [comment]
Looks good now. Could link to https://doc.rust-lang.org/book/ch10-00-generics.html maybe.

**Comment by SleeplessByte** [comment]
Close-reopened to make CI run.

**Review by SleeplessByte — APPROVED** [review]

**Pull request merged**